tasmota/support_device_groups.ino

```diff
@@ -0,0 +1,715 @@
+uint16_t outgoing_sequence = 0;
+bool initialized = false;
+bool building_status_message = false;
+bool processing_remote_device_message = false;
+bool waiting_for_acks;
+bool udp_was_connected = false;
+
+void DeviceGroupsInit(void)
+{
+  /*
+    Initialize the device information for each device group. The group name is the MQTT group topic.
+  */
+  if (Settings.flag.device_groups_enabled) {
+    device_groups = (struct device_group *)calloc(device_group_count, sizeof(struct device_group));
+    if (device_groups == nullptr) {
+      AddLog_P2(LOG_LEVEL_ERROR, "DGR: error allocating %u-element device group array", device_group_count);
+      Settings.flag.device_groups_enabled = false;
+      return;
+    }
+
+    for (uint32_t device_group_index = 0; device_group_index < device_group_count; device_group_index++) {
+      struct device_group * device_group = &device_groups[device_group_index];
+      strcpy(device_group->group_name, SettingsText((device_group_index == 0 ? SET_MQTT_GRP_TOPIC : SET_MQTT_GRP_TOPIC2 + device_group_index - 1)));
+      device_group->message_header_length = sprintf(device_group->message, "%s%s HTTP/1.1\n\n", kDeviceGroupMessage, device_group->group_name);
+      device_group->last_full_status_sequence = -1;
+    }
+
+    device_groups[0].local = true;
+
+    // If both in and out shared items masks are 0, assume they're unitialized and initialize them.
+    if (!Settings.device_group_share_in && !Settings.device_group_share_out) {
+      Settings.device_group_share_in = Settings.device_group_share_out = 0xff;
+    }
+
+    initialized = true;
+  }
+}
+
+char * IPAddressToString(const IPAddress& ip_address)
+{
+  static char buffer[16];
+  sprintf(buffer, "%u.%u.%u.%u", ip_address[0], ip_address[1], ip_address[2], ip_address[3]);
+  return buffer;
+}
+
+// Return true if we're configured to share the specified item.
+bool DeviceGroupItemShared(bool incoming, uint8_t item)
+{
+  uint8_t mask = 0;
+  switch (item) {
+    case DGR_ITEM_LIGHT_BRI:
+      mask = DGR_SHARE_LIGHT_BRI;
+      break;
+    case DGR_ITEM_POWER:
+      mask = DGR_SHARE_POWER;
+      break;
+    case DGR_ITEM_LIGHT_SCHEME:
+      mask = DGR_SHARE_LIGHT_SCHEME;
+      break;
+    case DGR_ITEM_LIGHT_FIXED_COLOR:
+    case DGR_ITEM_LIGHT_CHANNELS:
+      mask = DGR_SHARE_LIGHT_COLOR;
+      break;
+    case DGR_ITEM_LIGHT_FADE:
+    case DGR_ITEM_LIGHT_SPEED:
+      mask = DGR_SHARE_LIGHT_FADE;
+      break;
+    case DGR_ITEM_BRI_MIN:
+      mask = DGR_SHARE_BRI_MIN;
+      break;
+  }
+  return (!mask || ((incoming ? Settings.device_group_share_in : Settings.device_group_share_out) & mask));
+}
+
+void SendDeviceGroupPacket(IPAddress ip, char * packet, int len, const char * label)
+{
+  for (int attempt = 1; attempt <= 5; attempt++) {
+    if (PortUdp.beginPacket(ip, 1900)) {
+      PortUdp.write(packet, len);
+      if (PortUdp.endPacket()) return;
+    }
+    delay(10);
+  }
+  AddLog_P2(LOG_LEVEL_ERROR, "DGR: error sending %s packet", label);
```

Comment:
Please use PSTR to place string literals in Flash:

AddLog_P2(LOG_LEVEL_ERROR, PSTR("DGR: error sending %s packet"), label);